private/bufpkg/bufimage/bufimagemodify/bufimagemodifyv2/bufimagemodifyv2.go

```diff
@@ -15,49 +15,154 @@
 type Sweeper interface {
+	// Sweep removes SourceCodeInfo_Locations.
 	Sweep() error
 }
 
+// MarkSweeper marks SourceCodeInfo_Location indices and sweeps source code info
+// to remove marked SourceCodeInfo_Locations.
 type MarkSweeper interface {
 	Marker
 	Sweeper
 }
 
+// NewMarkSweeper returns a MarkSweeper.
 func NewMarkSweeper(image bufimage.Image) MarkSweeper {
-	return nil
+	return newMarkSweeper(image)
+}
+
+// Override describes how to modify a file option, and
+// may be passed to ModifyXYZ.
+type Override interface {
+	override()
+}
+
+type PrefixOverride interface {
+	Override
+	get() string
+	prefixOverride()
+}
+
+// NewPrefixOverride returns an override on prefix.
+func NewPrefixOverride(prefix string) PrefixOverride {
+	return newPrefixOverride(prefix)
+}
+
+type SuffixOverride interface {
+	Override
+	get() string
+	suffixOverride()
+}
+
+// NewSuffixOverride returns an override on suffix.
+func NewSuffixOverride(suffix string) SuffixOverride {
+	return newSuffixOverride(suffix)
+}
+
+type PrefixSuffixOverride interface {
+	Override
+	prefixSuffixOverride()
+}
+
+// NewPrefixSuffixOverride returns an override on both prefix and suffix.
+func NewPrefixSuffixOverride(prefixOverride PrefixOverride, suffixOverride SuffixOverride) PrefixSuffixOverride {
+	return newPrefixSuffixOverride(prefixOverride.get(), suffixOverride.get())
 }
 
+type ValueOverride interface {
+	Override
+	valueOverride()
+}
+
+// NewValueOverride returns a new override on value.
+func NewValueOverride[T string | bool | descriptorpb.FileOptions_OptimizeMode](val T) ValueOverride {
+	return newValueOverride[T](val)
+}
+
+// ModifyOption is an option for ModifyXYZ.
+type ModifyOption func(*modifyOptions)
+
+// ModifyWithOverride modifies an option with override.
+func ModifyWithOverride(override Override) (ModifyOption, error) {
+	if override == nil {
+		return nil, errors.New("override must not be nil")
+	}
+	return func(options *modifyOptions) {
+		options.override = override
+	}, nil
+}
+
+type modifyOptions struct {
+	override Override
+}
+
+// ModifyJavaPackage modifies java_package.
 func ModifyJavaPackage(
 	marker Marker,
 	imageFile bufimage.ImageFile,
-	prefix string,
+	modifyJavaPackageOptions ...ModifyOption,
 ) error {
+	options := &modifyOptions{}
+	for _, option := range modifyJavaPackageOptions {
+		option(options)
+	}
+	var javaPackageValue string
+	switch t := options.override.(type) {
+	case prefixOverride:
+		javaPackageValue = getJavaPackageValue(imageFile, t.get(), internal.DefaultJavaPackageSuffix)
+	case suffixOverride:
+		return errors.New("cannot modify java_package with a suffix but without a prefix")
+	case prefixSuffixOverride:
+		javaPackageValue = getJavaPackageValue(imageFile, t.prefix, t.suffix)
+	case valueOverride[string]:
+		javaPackageValue = t.get()
+	case nil:
+		javaPackageValue = getJavaPackageValue(imageFile, internal.DefaultJavaPackagePrefix, internal.DefaultJavaPackageSuffix)
```

Comment:
> The nil case shouldn't be valid. All overrides should have a value or prefix - if you just want the default value for managed mode, you don't set an override. Same comment whole way down

Now `ModifyXYZ` has option parameters that specify an override. Now `nil` means user of this package did not pass an option of override.